atc/db/versions_db.go

```diff
@@ -623,30 +590,73 @@ func (versions VersionsDB) migrateSingle(buildID int) (string, error) {
 					SELECT build_id, resource_id, json_agg(version_md5) AS v
 					FROM (
 						(
-							SELECT build_id, resource_id, version_md5 FROM build_resource_config_version_outputs o WHERE o.build_id = $1
+							SELECT build_id, resource_id, version_md5
+							FROM build_resource_config_version_outputs o
+							WHERE o.build_id = $1
 						)
 						UNION ALL
 						(
-							SELECT build_id, resource_id, version_md5 FROM build_resource_config_version_inputs i WHERE i.build_id = $1
+							SELECT build_id, resource_id, version_md5
+							FROM build_resource_config_version_inputs i
+							WHERE i.build_id = $1
 						)
 				) AS agg GROUP BY build_id, resource_id) sp ON sp.build_id = b.id
 				WHERE b.id = $1
 				GROUP BY b.id, b.job_id, b.rerun_of
-			) ON CONFLICT (build_id) DO UPDATE SET outputs = EXCLUDED.outputs RETURNING outputs`, buildID).
+			)
+			ON CONFLICT (build_id) DO UPDATE SET outputs = EXCLUDED.outputs
+			RETURNING outputs
+		`, buildID).
 		Scan(&outputs)
 	if err != nil {
+		tracing.End(span, err)
 		return "", err
 	}
 
+	span.AddEvent(ctx, "build migrated")
+
 	return outputs, nil
 }
 
+type BuildCursor struct {
+	ID      int
+	RerunOf sql.NullInt64
+}
+
+func (cursor BuildCursor) OlderBuilds(idCol string) sq.Sqlizer {
+	if cursor.RerunOf.Valid {
+		return sq.Or{
+			sq.Expr("COALESCE(rerun_of, "+idCol+") < ?", cursor.RerunOf.Int64),
```

Comment:
I think this won't catch the original build of a rerun (ex, if there are builds 5.1 and 5, it won't grab build 5 if the cursor is 5.1). Do we care about grabbing all the builds that are all reruns of each other? Since they all have the same input versions? Maybe we only care about grabbing latest rerun of a build?

Reply:
> I think this won't catch the original build of a rerun (ex, if there are builds 5.1 and 5, it won't grab build 5 if the cursor is 5.1).

Good catch - I'll add a test for this!

> Do we care about grabbing all the builds that are all reruns of each other? Since they all have the same input versions? Maybe we only care about grabbing latest rerun of a build?

Thought about this - I think we do, because while they all have the same inputs they might end up having different outputs.